CIPs/cip-TBC.md

```diff
@@ -0,0 +1,91 @@
+---
+cip: TBC
+title: Ed25519 precompile
+author: Joe Bowman <joe@chorus.one>
+discussions-to: https://github.com/celo-org/celo-proposals/issues/91
+status: Draft
+type: Standards
+category: Ring 0
+created: 2020-11-18
+license: Apache 2.0
+---
+
+### Terminology
+
+The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to be interpreted as described in 
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html).
+
+
+## Simple Summary
+
+Support performant and cheap verification of Ed25519 cryptographic signatures in smart contracts in general by adding a precompiled contract for Ed25519 signature verification to the EVM.
+
+## Abstract
+
+Verification of Ed25519 cryptographic signatures is possible in EVM bytecode. However, the gas cost is very high, and computationally expensive, as such tight, wide word operations intensive code as required for Ed25519 is not a good fit for the EVM bytecode model.
+
+The addition of a native compiled function, in a precompiled contract, to the EVM solves both cost and performance problems.
+
+## Motivation
+
+Ed25519 (that is, EdDSA using Curve25519) is an IETF recommendation (RFC7748) with some attractive properties:
+
+- Ed25519 is intended to operate at around the 128-bit security level.
+- EdDSA uses small public keys (32 octets) and signatures (64 octets) for Ed25519.
+- Ed25519 are designed so that fast, constant-time (timing attack resistant) and generally side-channel resistant implementations are easier to produce.
+
+With an Ed25519 precompile, we are able to verify Ed25519 signatures origination from external, Ed25519 proof-of-stake based blockchains within Celo EVM smart contracts, allowing for proof-based trustless bridges between Celo and other blockchains.
+
+## Specification
+
+If block.number >= FORK_HEIGHT (TBC), add a precompiled contract for Ed25519 signature verification (`ED25519VFY`).
+
+The proposal adds a new precompiled function `ED25519VFY` with the following input and output.
+
+`ED25519VFY` takes as input 128 octets:
+
+- message: the 32-octet message that was signed
+- public key: the 32-octet Ed25519 public key of the signer
+- signature: the 64-octet Ed25519 signature
+
+`ED25519VFY` returns as output 4 octets:
+
+- `0x00000000` if signature is valid
+- any non-zero value indicates a signature verification failure
```

Comment:
Will there be an initially defined error code? Is there any way for this precompile to revert?